File: 택시 합승 요금.py
INF = int(1e9)

def solution(n, s, a, b, fares):
    length = len(fares) #간선의 개수
    graph = [[INF] * (n+1) for _ in range(n+1)]
    
    for q in range(1,n+1):
        for w in range(1,n+1):
            if q==w:
                graph[q][w] = 0
    
    for q,w,e in fares:
        graph[q][w] = e
        graph[w][q] = e

    for k in range(1,n+1):
        for q in range(1,n+1):
            for w in range(1,n+1):
                if(graph[q][w] > graph[q][k]+graph[k][w]):
                    graph[q][w] = graph[q][k]+graph[k][w]
    
    answer = int(INF)
    for i in range(1,n+1):
        answer =min(answer,graph[s][i]+graph[i][a]+graph[i][b])
        
    return answer

File: 택시 합승 요금.java
import java.lang.*;
class Solution {
    public int solution(int n, int s, int a, int b, int[][] fares) {
        int INF = 4000001;
        
        int[][] graph = new int[n+1][n+1];
        for(int i=1;i<n+1;i++){
            for(int j=1;j<n+1;j++){
                graph[i][j] = INF;
            }
        }
        for(int i=1;i<n+1;i++){
            for(int j=1;j<n+1;j++){
                if (i == j){
                    graph[i][j] = 0;
                }
            }
        }
        for(int i=0;i<fares.length;i++){
            graph[fares[i][0]][fares[i][1]] = fares[i][2];
            graph[fares[i][1]][fares[i][0]] = fares[i][2];
        }
        
        for(int k=1;k<n+1;k++){
            for(int i=1;i<n+1;i++){
                for(int j=1;j<n+1;j++){
                    graph[i][j] = Math.min(graph[i][j],graph[i][k]+graph[k][j]);
                }
            }
        }
        int answer = INF;
        for(int i=1;i<n+1;i++){
            answer = Math.min(answer,graph[s][i]+graph[i][a]+graph[i][b]);
        }
        
        return answer;
    }
}
